### PR TITLE
[jsonwebtoken] change decode object signature to be less restrictive

### PR DESCRIPTION
**Pull request by brikou**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If adding a new definition:
- [ ] The package does not provide its own types, and you can not add them.
- [ ] If this is for an NPM package, match the name. If not, do not conflict with the name of an NPM package.
- [ ] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [ ] `tslint.json` should be present, and `tsconfig.json` should have `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

If changing an existing definition:
- [ ] Provide a URL to documentation or source code which provides context for the suggested changes: <<url here>>
- [ ] Increase the version number in the header if appropriate.
- [ ] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

If removing a declaration:
- [ ] If a package was never on DefinitelyTyped, you don't need to do anything. (If you wrote a package and provided types, you don't need to register it with us.)
- [ ] Delete the package's directory.
- [ ] Add it to `notNeededPackages.json`.

### CODE REVIEW

File: types/jsonwebtoken/index.d.ts [hunk 1]
```diff
@@ -166,4 +166,4 @@ declare function verify(
 declare function decode(
     token: string,
     options?: DecodeOptions,
-): null | object | string;
+): null | { [key: string]: any } | string;
```
Comment (aozgaa) [inline types/jsonwebtoken/index.d.ts]: Why change from `object`?

**Review by paulvanbrenk — CHANGES_REQUESTED** [review]
I think the correct implementation would import @types\jws and use those types, which is what the implementation appears to do.

File: types/jsonwebtoken/index.d.ts (hunk 1 above)
Comment (brikou) [reply types/jsonwebtoken/index.d.ts]: Using `object` is equivalent to `{}`, using `{ [key: string]: any }` describe the general shape of the object and so makes it less restrictive.

**Comment by brikou** [comment]
@paulvanbrenk I don't understand your comment can you elaborate?

**Comment by paulvanbrenk** [comment]
@brikou the implementation of jwt seems to depend on jws, so I suggested using the definition file for that component so specify the return types of e.g. decode.


https://github.com/auth0/node-jsonwebtoken/blob/master/decode.js

Though arguably that change may be more than you bargained for, or I'm completely wrong.

File: types/jsonwebtoken/index.d.ts (hunk 1 above)
Comment (mhegazy) [reply types/jsonwebtoken/index.d.ts]: > Using `object` is equivalent to `{}`, using `{ [key: string]: any }` describe the general shape of the object and so makes it less restrictive.
  
you do not need the `| string` then.. it does match `{ [key: string]: any } `

Comment (andy-ms) [reply types/jsonwebtoken/index.d.ts]: By less restrictive, you mean it's not type-checked? Can't you leave the declaration as `object` nd cast it to the type you expect?

**Pull request merged**